src/librustc_ast_passes/ast_validation.rs

```diff
@@ -826,28 +826,36 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
                 if let Unsafe::Yes(span) = unsafety {
                     struct_span_err!(
                         self.session,
-                        item.span,
+                        vec![span, self_ty.span],
```

Comment:
I think I'd prefer not altering the primary span here too save for using `sm.def_span(item.span)`.

Reply:
I don't feel like `^^^---^^^` is problematic. Do we have any evidence to support this? Anyways, I guess I'm OK with changing the primary span for now, but I'd still like a `span_label` for the `!` bit.

> There's also the intention to move away from the need to use `SourceMap::def_span` as it is a hack, but that is a less pressing matter for me than reducing how much we highlight in any given error.

Feels like a pretty reliable hack to me in terms of getting good diagnostics output, so idk why we want to get rid of it.